nifi-nar-bundles/nifi-atlas-bundle/nifi-atlas-reporting-task/src/main/java/org/apache/nifi/atlas/provenance/StandardAnalysisContext.java

```diff
@@ -0,0 +1,136 @@
+import org.apache.nifi.provenance.ProvenanceEventRecord;
+import org.apache.nifi.provenance.ProvenanceRepository;
+import org.apache.nifi.provenance.lineage.ComputeLineageResult;
+import org.apache.nifi.provenance.lineage.ComputeLineageSubmission;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+public class StandardAnalysisContext implements AnalysisContext {
+
+    private final Logger logger = LoggerFactory.getLogger(StandardAnalysisContext.class);
+    private final NiFiFlow nifiFlow;
+    private final ClusterResolver clusterResolver;
+    private final ProvenanceRepository provenanceRepository;
+
+    public StandardAnalysisContext(NiFiFlow nifiFlow, ClusterResolver clusterResolver,
+                                   ProvenanceRepository provenanceRepository) {
+        this.nifiFlow = nifiFlow;
+        this.clusterResolver = clusterResolver;
+        this.provenanceRepository = provenanceRepository;
+    }
+
+    @Override
+    public List<ConnectionStatus> findConnectionTo(String componentId) {
+        return nifiFlow.getIncomingRelationShips(componentId);
+    }
+
+    @Override
+    public List<ConnectionStatus> findConnectionFrom(String componentId) {
+        return nifiFlow.getOutgoingRelationShips(componentId);
+    }
+
+    @Override
+    public String getNiFiClusterName() {
+        return nifiFlow.getClusterName();
+    }
+
+    @Override
+    public ClusterResolver getClusterResolver() {
+        return clusterResolver;
+    }
+
+    private ComputeLineageResult getLineageResult(long eventId, ComputeLineageSubmission submission) {
+        final ComputeLineageResult result = submission.getResult();
+        try {
+            if (result.awaitCompletion(10, TimeUnit.SECONDS)) {
+                return result;
+            }
+            logger.warn("Lineage query for {} timed out.", new Object[]{eventId});
+        } catch (InterruptedException e) {
+            logger.warn("Lineage query for {} was interrupted due to {}.", new Object[]{eventId, e}, e);
+        } finally {
+            submission.cancel();
+        }
+
+        return null;
+    }
+
+    @Override
+    public ComputeLineageResult queryLineage(long eventId) {
+        final ComputeLineageSubmission submission = provenanceRepository.submitLineageComputation(eventId, NIFI_USER);
+        return getLineageResult(eventId, submission);
+    }
+
+    public ComputeLineageResult findParents(long eventId) {
+        final ComputeLineageSubmission submission = provenanceRepository.submitExpandParents(eventId, NIFI_USER);
+        return getLineageResult(eventId, submission);
+    }
+
+    // NOTE: This user is required to avoid NullPointerException at PersistentProvenanceRepository.submitLineageComputation
+    private static final QueryNiFiUser NIFI_USER = new QueryNiFiUser();
```

Comment:
Using this approach means that any NiFi that is run securely will not be able to make use of this Atlas integration. I think the more correct solution is to update the Provenance Repository so that we can run a query without supplying a user. In the mean time, we could perhaps get away with allowing the user to configure the username to run the queries as? Then they can simply create an "Atlas Integration" user or choose their own user, etc.

Reply:
Actually this user can be anything. Since NiFI authentication and authorization is done at the web request layer, when the provenance API is called directly, no authentication nor authorization take place. I confirmed that provenance query work in secured NiFi environment. Is that approach fine?

Reply:
@ijokarumawak The authentication is done at the web request layer and authorization for making a provenance query is done there, but the event-level authorization is not. That is done in the repository itself. So in a secure environment, the lineage that is returned should consist of "UNKNOWN" event types. If that is not the behavior you are seeing, then I believe we have a bug somewhere.

Reply:
I see. What I am seeing is different from your description. I will try to find where the event level auth is implemented.

Reply:
@ijokarumawak I think I understand now what is going on. In a secured environment, you would indeed get back all UNKNOWN event types, because that user would not be authorized to see the events. But even the UNKNOWN event would include the ID of the event. It looks like you're not actually using the other info from the lineage but rather just using the Event ID to lookup the actual Event (doing so without a user). So all of this seems to me now :) Sorry about all of the confusion - and thanks for the clarifications!